frontend/app/hooks/use-tab-id.ts

```diff
@@ -0,0 +1,107 @@
+import { useEffect, useSyncExternalStore } from 'react';
+
+import { useLocation, useNavigate } from 'react-router';
+
+import { randomString } from '~/utils/string-utils';
+
+export const SEARCH_PARAM_KEY = 'tid';
+export const SESSION_STORAGE_KEY = 'tab-id';
+
+/**
+ * Generates a random tab identifier in the format `xx-0000`.
+ * This id will uniquely identify different tabs within the same browser instance.
+ */
+function generateRandomId(): string {
+  const prefix = randomString(2, 'abcdefghijklmnopqrstuvwxyz');
+  const suffix = randomString(4, '0123456789');
+  return `${prefix}-${suffix}`;
+}
+
+/**
+ * Retrieves the current tab id from session storage or generates a new one if not found.
+ * Ensures the tab id persists across page reloads within the same tab.
+ */
+function getSnapshot(sessionStorageKey: string): string {
+  const id = window.sessionStorage.getItem(sessionStorageKey) ?? generateRandomId();
+  window.sessionStorage.setItem(sessionStorageKey, id); // store the id to persist it across reloads
+  return id;
+}
+
+/**
+ * Subscribes to `storage` events to detect changes in session storage.
+ * If the tab id in session storage changes (e.g., via tampering),
+ * it triggers a re-render by calling the provided callback.
+ */
+function subscribe(sessionStorageKey: string, callback: () => void): () => void {
+  const handler = ({ key }: StorageEvent): void => {
+    if (key === sessionStorageKey) {
+      callback(); // trigger a state update only if the session storage key changes
+    }
+  };
+
+  window.addEventListener('storage', handler);
+  return () => window.removeEventListener('storage', handler);
+}
+
+type Options = {
+  /**
+   * Whether to automatically update the URL with the tab id as a query parameter.
+   * Defaults to `true`.
+   */
+  navigate?: false;
+  /**
+   * The query parameter key used for storing the tab id in the URL.
+   * Defaults to `'tid'`.
+   */
+  searchParamKey?: string;
+  /**
+   * The session storage key used for persisting the tab id.
+   * Defaults to `'tab-id'`.
+   */
+  sessionStorageKey?: string;
+};
+
+/**
+ * React hook that provides a unique, persistent identifier for the current browser tab.
+ *
+ * - The id persists across page reloads but resets when the tab is closed.
+ * - If `navigate` is enabled (default: `true`), the hook ensures the id is present in the URL.
```

Comment:
Why the tabId has to be present in the URL?

Reply:
The server can't access `window.sessionStorage` so it has to be added to the URL so it can be used server side when storing stuff in session there.

Reply:
The idea is that the data will be stored server side, but keyed according to the tabid, so we can have multiple tabs with different multi-page flows without them each clobbering one another.

Reply:
Ok. I will have to see how it's gonna be used later, I guess.

Reply:
It's very similar to CDCP, except this one will work (CDCP's implementation is a bit buggy).